distributed/shuffle/_rechunk.py

```diff
@@ -340,16 +350,17 @@ def partial_concatenate(
             axis[index] for index, axis in zip(old_global_index, x.chunks)
         )
         if _slicing_is_necessary(ndslice, original_shape):
-            rec_cat_arg[old_partial_index] = (slice_name,) + old_global_index
-            dsk[(slice_name,) + old_global_index] = (
+            key = (slice_name,) + ndpartial.ix + old_global_index
+            rec_cat_arg[old_partial_index] = key
+            dsk[key] = (
```

Comment:
So IIUC, this is grouping tasks more accurately (and "sparsifying") by moving the "point" index (represented by `ndpartial.ix`) from the input token to the arguments to `getitem`. This way, tasks that access the same bit(s) share inputs/outputs, but we don't introduce all-to-all coupling?

Reply:
See https://github.com/dask/distributed/pull/8655#discussion_r1616794792. Essentially, we just move information around within the task key. Specifically, we move the information about the partials out of the group so that we end up with a single group for the entire rechunk and move it into the remainder of the key.